integrations/sonarqube/client.py

```diff
@@ -468,6 +472,39 @@ async def get_all_sonarqube_analyses(
             )
             yield analysis_data
 
+    async def _get_all_portfolios(self) -> list[dict[str, Any]]:
+        logger.info(f"Fetching all portfolios in organization: {self.organization_id}")
+        response = await self.send_api_request(endpoint=Endpoints.PORTFOLIOS)
+        return response.get("views", [])
+
+    async def get_portfolio_details(self, portfolio_key: str) -> dict[str, Any]:
+        logger.info(f"Fetching portfolio details for: {portfolio_key}")
+        response = await self.send_api_request(
+            endpoint=Endpoints.PORTFOLIO_DETAILS,
+            query_params={"key": portfolio_key},
+        )
+        return response
+
+    def _get_portfolio_subportfolios(
+        self, portfolio: dict[str, Any]
+    ) -> list[dict[str, Any]]:
+        logger.info(f"Fetching subportfolios for: {portfolio['key']}")
+        subportfolios = portfolio.get("subViews", []) or []
+        all_portfolios = []
+        for subportfolio in subportfolios:
+            if subportfolio.get("qualifier") in PORTFOLIO_VIEW_QUALIFIERS:
+                all_portfolios.append(subportfolio)
+            all_portfolios.extend(self._get_portfolio_subportfolios(subportfolio))
+        return all_portfolios
+
+    async def get_all_portfolios(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        portfolios = await self._get_all_portfolios()
+
+        for portfolio in portfolios:
+            portfolio_data = await self.get_portfolio_details(portfolio["key"])
+            yield [portfolio_data]
+            yield self._get_portfolio_subportfolios(portfolio_data)
```

Comment:
can we add some try catch here so that in the event there's an unexpected error, we can catch it?

Reply:
also, can we fetch all the portfolio details at once using async gather, and then take our time to extra it's sub portfolio data recursively? 

```
    portfolio_keys = [portfolio["key"] for portfolio in portfolios]
    portfolio_details = await asyncio.gather(
        *[self.get_portfolio_details(key) for key in portfolio_keys]
    )
```
for each portfolio details list
  yield portfolio
  yield extract sub portfolio( portfolio)